src/synthesizer/pipeline/pipeline.py

```diff
@@ -0,0 +1,1801 @@
+                if g.stars is not None:
+                    if not hasattr(g.stars, "images_psf_fnu"):
+                        g.stars.images_psf_fnu = {}
+                    g.stars.images_psf_fnu.setdefault(inst.label, {})
+                    for spec_type, imgs in g.stars.images_fnu[
+                        inst.label
+                    ].items():
+                        g.stars.images_psf_fnu[inst.label][spec_type] = (
+                            imgs.apply_psfs(
+                                inst.psfs,
+                            )
+                        )
+
+                        # Unpack the image arrays onto the Pipeline object
+                        for f, img in g.stars.images_psf_fnu[inst.label][
+                            spec_type
+                        ].items():
+                            self.images_flux_psf["Stars"].setdefault(
+                                spec_type, {}
+                            ).setdefault(f, []).append(img.arr * img.units)
+
+                # Apply PSFs to the black hole level images
+                if g.black_holes is not None:
+                    if not hasattr(g.black_holes, "images_psf_fnu"):
+                        g.black_holes.images_psf_fnu = {}
+                    g.black_holes.images_psf_fnu.setdefault(inst.label, {})
+                    for spec_type, imgs in g.black_holes.images_fnu[
+                        inst.label
+                    ].items():
+                        g.black_holes.images_psf_fnu[inst.label][spec_type] = (
+                            imgs.apply_psfs(
+                                inst.psfs,
+                            )
+                        )
+
+                        # Unpack the image arrays onto the Pipeline object
+                        for f, img in g.black_holes.images_psf_fnu[inst.label][
+                            spec_type
+                        ].items():
+                            self.images_flux_psf["BlackHole"].setdefault(
+                                spec_type, {}
+                            ).setdefault(f, []).append(img.arr * img.units)
+
+        # Convert the lists of images to unyt arrays
+        for spec_type, imgs in self.images_flux_psf["Galaxy"].items():
+            for f, img in imgs.items():
+                self.images_flux_psf["Galaxy"][spec_type][f] = unyt_array(img)
+        for spec_type, imgs in self.images_flux_psf["Stars"].items():
+            for f, img in imgs.items():
+                self.images_flux_psf["Stars"][spec_type][f] = unyt_array(img)
+        for spec_type, imgs in self.images_flux_psf["BlackHole"].items():
+            for f, img in imgs.items():
+                self.images_flux_psf["BlackHole"][spec_type][f] = unyt_array(
+                    img
+                )
+
+        # Done!
+        self._got_images_flux_psf = True
+        self._took(start, "Applying PSFs to flux images")
+
+    def get_lnu_data_cubes(self):
```

Comment:
Why the4 different naming convention here compared to the spectra and images?

Reply:
In short, no idea 😅 I had ignored them after adding the `NotImplemented` error.